### PR TITLE
Filter records by type

### PR DESCRIPTION
**Pull request by j0s**
Add API to only get events of a certain type when getting events by sequence ID.

### CODE REVIEW

File: eventsource/repository.go [hunk 1]
```diff
@@ -184,28 +189,37 @@ func (repo repository) Load(ctx context.Context, aggregateID string, aggr Aggreg
...
+	var records []Record
+	if records, err = repo.store.LoadBySequenceID(ctx, sequenceID); err != nil {
+		return
+	}
+	return unmarshalRecords(repo.serializer, records)
+}
+
+func (repo repository) GetEventsBySequenceIDAndType(ctx context.Context, sequenceID string, eventType Event) (events []Event, err error) {
+	var records []Record
+	if records, err = repo.store.LoadBySequenceIDAndType(ctx, sequenceID, reflect.TypeOf(eventType).Name()); err != nil {
```
Comment (stroem) [inline eventsource/repository.go]: `reflect.TypeOf(eventType).Name()` will not work if eventType is of type pointer

```
func getTypeOfValue(input interface{}) reflect.Type {
	value := reflect.TypeOf(input)
	if value.Kind() == reflect.Ptr {
		value = value.Elem()
	}
	return value
}
```

`getTypeOfValue(eventType).Name()` this will give you what you want, as long as the eventType isn't a double pointer :D

The function above already exists in the serilizer. Would be nice if you updated `SaveTransaction` that has the same fault

File: eventsource/stores/memorystore/store.go [hunk 2]
```diff
@@ -32,6 +32,26 @@ func (mem *store) LoadBySequenceID(_ context.Context, sequenceID string) (record
 			if row.SequenceID > sequenceID {
 				records = append(records, row)
 			}
+			if len(records) >= 1000 {
```
Comment (stroem) [inline eventsource/stores/memorystore/store.go]: This will make this function pretty much "useless" without pagination, why is this needed?

File: eventsource/stores/memorystore/store.go [hunk 3]
```diff
@@ -32,6 +32,26 @@ func (mem *store) LoadBySequenceID(_ context.Context, sequenceID string) (record
...
+		}
+	}
+	sort.Slice(records, func(i, j int) bool {
+		return records[i].SequenceID < records[j].SequenceID
+	})
+	return
+}
+
+func (mem *store) LoadBySequenceIDAndType(_ context.Context, sequenceID string, eventType string) (records []eventsource.Record, err error) {
+	for _, aggregate := range mem.Data {
```
Comment (stroem) [inline eventsource/stores/memorystore/store.go]: Duplicated code, maybe fine?

File: eventsource/stores/sqlstore/store.go [hunk 4]
```diff
@@ -14,10 +14,11 @@ type store struct {
 }
 
 const (
-	saveSQL = "INSERT INTO %s (aggregate_id, sequence_id, created_at, user_id, type, data) VALUES ($1, $2, $3, $4, $5, $6)"
-	loadAggregateSQL = "SELECT aggregate_id, sequence_id, created_at, user_id, type, data FROM %s WHERE aggregate_id = $1 ORDER BY sequence_id ASC LIMIT 100000"
-	loadBySequenceIDSQL = "SELECT aggregate_id, sequence_id, created_at, user_id, type, data FROM %s WHERE sequence_id > $1 ORDER BY sequence_id ASC LIMIT 100000"
-	loadByTimestampSQL = "SELECT aggregate_id, sequence_id, created_at, user_id, type, data FROM %s WHERE timestamp > $1 ORDER BY timestamp ASC LIMIT 100000"
+	saveSQL                    = "INSERT INTO %s (aggregate_id, sequence_id, created_at, user_id, type, data) VALUES ($1, $2, $3, $4, $5, $6)"
+	loadAggregateSQL           = "SELECT aggregate_id, sequence_id, created_at, user_id, type, data FROM %s WHERE aggregate_id = $1 ORDER BY sequence_id ASC LIMIT 100000"
+	loadBySequenceIDSQL        = "SELECT aggregate_id, sequence_id, created_at, user_id, type, data FROM %s WHERE sequence_id > $1 ORDER BY sequence_id ASC LIMIT 1000"
```
Comment (stroem) [inline eventsource/stores/sqlstore/store.go]: Why the limitation?

File: eventsource/stores/sqlstore/store.go [hunk 5]
```diff
@@ -63,6 +64,10 @@ func (store *store) LoadBySequenceID(ctx context.Context, sequenceID string) (re
 	return store.createRecords(ctx, loadBySequenceIDSQL, sequenceID)
 }
 
+func (store *store) LoadBySequenceIDAndType(ctx context.Context, sequenceID string, eventType string) (records []eventsource.Record, err error) {
+	return store.createRecords(ctx, loadBySequenceIDAndTypeSQL, sequenceID, eventType)
```
Comment (stroem) [inline eventsource/stores/sqlstore/store.go]: `createRecords`? more like `fetchRecords`

File: eventsource/stores/sqlstore/store_test.go [hunk 6]
```diff
@@ -144,7 +163,7 @@ func Test_SQLStoreE2E(t *testing.T) {
 	}
 	defer db.Close()
 
-	tmpTableName := "hejsan"//strings.Replace(uuid.New().String(), "-", "", -1)
+	tmpTableName := "hejsan" //strings.Replace(uuid.New().String(), "-", "", -1)
```
Comment (stroem) [inline eventsource/stores/sqlstore/store_test.go]: Something you forgot to undo?

File: eventsource/stores/sqlstore/store_test.go [hunk 7]
```diff
@@ -197,5 +216,5 @@ CREATE TABLE %s (
 	assert.Nil(t, err, "Could not get events")
 	assert.Equal(t, 1, len(events))
 
-	db.Exec("DROP TABLE $1", tmpTableName)
+	db.Exec(fmt.Sprintf("DROP TABLE IF EXISTS %s", tmpTableName))
```
Comment (stroem) [inline eventsource/stores/sqlstore/store_test.go]: Shouldn't this run as a defer after table is created?

Comment (stroem) [inline eventsource/stores/sqlstore/store_test.go]: L178 should be `require.NoError(t, err...` 
It should not continue if the creation of the table wasn't succesful. It could be dangerous if the table that already exists got deleted

Comment (stroem) [inline eventsource/stores/sqlstore/store_test.go]: Handle errors from db.Exec

File: eventsource/repository_test.go [hunk 8]
```diff
@@ -94,6 +103,27 @@ func Test_RepoGetRecords(t *testing.T) {
 	assert.Equal(t, records[1].(*OtherEvent), otherEvent, true)
 }
 
+func Test_RepoGetRecordsByType(t *testing.T) {
+	storeMock, _, serializerMock, aggregatorMock := setupMocks()
+
+	history, baseEvent, _ := createMockDataForLoadAggregate()
+	otherEvent := &OtherEvent{BaseEvent: baseEvent, OtherEventField: 42}
+
+	ctx := context.Background()
```
Comment (stroem) [inline eventsource/repository_test.go]: Hmm.. Wondering if we should use context.TODO() for testing.
https://golang.org/src/context/context.go?s=7708:7727#L205

File: eventsource/repository.go (hunk 1 above)
Comment (j0s) [reply eventsource/repository.go]: Good catch. We could even handle double pointers if instead of `value = value.Elem()` we do `return getTypeOfValue(value.Elem())` :)

File: eventsource/stores/sqlstore/store.go (hunk 5 above)
Comment (j0s) [reply eventsource/stores/sqlstore/store.go]: True, will fix!

File: eventsource/stores/sqlstore/store_test.go (hunk 7 above)
Comment (j0s) [reply eventsource/stores/sqlstore/store_test.go]: Yes, that would probably be better

File: eventsource/stores/memorystore/store.go (hunk 2 above)
Comment (j0s) [reply eventsource/stores/memorystore/store.go]: The alarm status event table has > 200000 records. Unmarshalling them all is time consuming, and results in messages that are far too big to send over gRPC. Also, doing something meaningful with 200000 records in a lambda takes too much time. I realize that these limitations are mostly related to my specific use case - maybe the limit should be given as a parameter?

File: eventsource/stores/sqlstore/store.go (hunk 4 above)
Comment (j0s) [reply eventsource/stores/sqlstore/store.go]: See https://github.com/SKF/go-eventsource/pull/31/files/2fb075345283b77c2523648dc6fcfbf83cab0929#r250258077
